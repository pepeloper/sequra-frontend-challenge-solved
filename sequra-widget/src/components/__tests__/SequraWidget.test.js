import { mount } from '@vue/test-utils'
import { beforeEach } from 'vitest'
import SequraWidget from '../SequraWidget.ce.vue'

global.fetch = vi.fn()

function createFetchResponse(data, httpStatus = 200) {
  return {
    status: httpStatus,
    json: () => new Promise((resolve) => resolve(data)),
  }
}

describe('SequraWidget', () => {

  beforeEach(() => {
    global.fetch.mockReset()
    fetch.mockResolvedValue(createFetchResponse([
      {
        "instalment_count": 3,
        "instalment_amount": { "string": "133,33 €" },
        "instalment_fee": { "string": "5 €" },
        "instalment_total": { "string": "138,33 €" },
      },
      {
        "instalment_count": 6,
        "instalment_amount": { "string": "66,66 €" },
        "instalment_fee": { "string": "5 €" },
        "instalment_total": { "string": "71,66 €" },
      },
    ]))
  })

  test('mount component', () => {
    expect(SequraWidget).toBeTruthy()

    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: 40000,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(fetch.calls.length).toBe(2)
  })

  test('calculate the credit agreement on load', () => {
    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: 40000,
      },
    })

    expect(fetch).toHaveBeenCalledWith('http://localhost:8080/credit_agreements?totalWithTax=40000')
  })

  test('send event on load', () => {
    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: 55000,
      },
    })

    expect(fetch).toHaveBeenCalledWith('http://localhost:8080/events', {
      method: 'post',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ type: 'simulatorInstalmentTotalInitialized', totalWithTax: 55000, context: 'checkoutWidget' }),
    })
  })

  test('re-calculate credit agreement on prop change', async () => {
    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: 55000,
      },
    })

    await wrapper.setProps({ totalWithTax: 20000 })

    expect(fetch).toHaveBeenCalledWith('http://localhost:8080/credit_agreements?totalWithTax=20000')
  })

  test('send event on prop change', async () => {
    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: 55000,
      },
    })

    await wrapper.setProps({ totalWithTax: 20000 })

    expect(fetch).toHaveBeenCalledWith('http://localhost:8080/events', {
      method: 'post',
      body: JSON.stringify({ type: 'simulatorInstalmentTotalChanged', totalWithTax: 20000, context: 'checkoutWidget' }),
      headers: {
        'Content-Type': 'application/json'
      },
    })
  })

  test('open modal set to true by clicking on more info', async () => {
    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: 40000,
      },
    })

    const button = await wrapper.find('.sequra__widget__modal__button')
    button.trigger('click')

    expect(wrapper.vm.modalIsOpened).toBeTruthy()
  })

  test('it doesnt render or make api calls if total with tax is not set with a number', () => {
    expect(SequraWidget).toBeTruthy()

    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: null,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.vm.widgetError).toBeTruthy()
    expect(fetch.calls.length).toBe(0)
  })

  test('set error properly if totalWithTax is changed to not a number', async () => {
    const wrapper = mount(SequraWidget, {
      props: {
        totalWithTax: 40000,
      },
    })

    expect(wrapper.vm.widgetError).toBeFalsy()

    await wrapper.setProps({ totalWithTax: null })

    expect(wrapper.vm.widgetError).toBeTruthy()
  })
})