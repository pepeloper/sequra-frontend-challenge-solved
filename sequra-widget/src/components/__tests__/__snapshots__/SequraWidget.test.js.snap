// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SequraWidget > it doesnt render or make api calls if total with tax is not set with a number 1`] = `"<!--v-if-->"`;

exports[`SequraWidget > mount component 1`] = `
"<div class=\\"sequra__widget\\">
  <div class=\\"sequra__widget__heading\\">
    <p>Págalo en</p><button class=\\"sequra__widget__modal__button\\" type=\\"button\\">más info</button>
  </div>
  <div class=\\"sequra__widget__content\\"><select name=\\"credit-agreement\\" id=\\"sequra__credit__agreement__selector\\"></select></div>
  <div class=\\"sequra__widget__modal__overlay\\" style=\\"display: none;\\">
    <div class=\\"sequra__widget__modal\\" role=\\"dialog\\" aria-labelledby=\\"sequra-widget__label\\" aria-hidden=\\"true\\">
      <div class=\\"sequra__widget__modal__header\\">
        <p id=\\"sequra-widget__label\\">Fracciona tu pago</p>
      </div>
      <div class=\\"sequra__widget__modal__body\\">
        <ol>
          <li>Eliges \\"Fracciona tu pago\\" al realizar tu pedido y pagas solo la primera cuota</li>
          <li>Recibes tu pedido</li>
          <li>El resto de pagos se cargarán automáticamente a tu tarjeta</li>
        </ol>
        <p>¡Así de simple!</p>
        <p> Además, en el importe mostrado ya se incluye la cuota única mensual de /mes por lo que no tendrás ninguna sorpresa. </p>
      </div>
    </div>
  </div>
</div>"
`;
